File: 백준/Silver/1931. 회의실 배정/회의실 배정.py
import sys
input = sys.stdin.readline

n = int(input())
meetings = [tuple(map(int, input().split())) for _ in range(n)]

meetings.sort(key=lambda x: (x[1], x[0])) # 종료 시각 기준으로 정렬하고, 종료 시각이 같으면 시작 시각 기준 정렬하기

count = 1
end_time = meetings[0][1]

for i in range(1, n):
    if meetings[i][0] >= end_time: # 앞 회의의 종료 시각보다 시작 시각이 늦은 회의가 나온 경우
        count += 1
        end_time = meetings[i][1]

print(count)

File: 백준/Silver/1931. 회의실 배정/회의실 배정.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        List<int[]> meetings = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            int start = sc.nextInt();
            int end = sc.nextInt();
            meetings.add(new int[] {start, end});
        }

        // 종료 시각 기준으로 정렬하기
        meetings.sort((a, b) -> {
            if (a[1] == b[1]) {
                return Integer.compare(a[0], b[0]);
            } else {
                return Integer.compare(a[1], b[1]);
            }
        });

        // 회의 선택하기
        int count = 1;
        int end_time = meetings.get(0)[1];

        for (int i = 1; i < n; i++) {
            if (meetings.get(i)[0] >= end_time) {
                count++;
                end_time = meetings.get(i)[1];
            }
        }

        System.out.println(count);
        sc.close();
    }
}

